src/network_service/sending_targets_cache.rs

```diff
@@ -0,0 +1,144 @@
+
+const MAX_RESENDS: u8 = 3;
+
+enum TargetState {
+    Sending(u8),
+    Failed(u8),
+    Sent,
+}
+
+impl TargetState {
+    pub fn is_complete(&self) -> bool {
+        match *self {
+            TargetState::Failed(x) => x > MAX_RESENDS,
+            TargetState::Sent => true,
+            TargetState::Sending(_) => false,
+        }
+    }
+
+    pub fn is_sending(&self) -> bool {
+        match *self {
+            TargetState::Failed(_) | TargetState::Sent => false,
+            TargetState::Sending(_) => true,
+        }
+    }
+}
+
+#[derive(Default)]
+pub struct SendingTargetsCache {
+    cache: HashMap<u64, Vec<(ConnectionInfo, TargetState)>>,
+}
+
+impl SendingTargetsCache {
+    pub fn insert_message(
+        &mut self,
+        msg_id: u64,
+        initial_targets: Vec<ConnectionInfo>,
+        dg_size: usize,
+    ) {
+        let targets = initial_targets
+            .into_iter()
+            .enumerate()
+            .map(|(idx, tgt_info)| {
+                (
+                    tgt_info,
+                    if idx < dg_size {
+                        TargetState::Sending(0)
+                    } else {
+                        TargetState::Failed(0)
+                    },
+                )
+            })
+            .collect();
+        let _ = self.cache.insert(msg_id, targets);
+    }
+
+    fn target_states(&self, msg_id: u64) -> impl Iterator<Item = &(ConnectionInfo, TargetState)> {
+        self.cache.get(&msg_id).into_iter().flatten()
+    }
+
+    fn target_states_mut(
+        &mut self,
+        msg_id: u64,
+    ) -> impl Iterator<Item = &mut (ConnectionInfo, TargetState)> {
+        self.cache.get_mut(&msg_id).into_iter().flatten()
+    }
+
+    fn fail_target(&mut self, msg_id: u64, target: SocketAddr) {
+        let _ = self
+            .target_states_mut(msg_id)
+            .find(|(info, _state)| info.peer_addr() == target)
+            .map(|(_info, state)| match *state {
+                TargetState::Failed(_) => {
+                    log_or_panic!(LogLevel::Error, "Got a failure from a failed target!");
+                }
+                TargetState::Sending(x) => {
+                    *state = TargetState::Failed(x + 1);
+                }
+                TargetState::Sent => {
+                    log_or_panic!(
+                        LogLevel::Error,
+                        "A target that should no longer fail - failed!"
+                    );
+                }
+            });
+    }
+
+    fn take_next_target(&mut self, msg_id: u64) -> Option<ConnectionInfo> {
+        self.target_states_mut(msg_id)
+            .filter(|(_info, state)| !state.is_complete())
+            .filter_map(|(info, state)| match state {
+                TargetState::Failed(x) => Some((info, *x, state)),
+                _ => None,
+            })
+            .min_by_key(|(_info, num, _state)| *num)
+            .map(|(info, num, state)| {
+                *state = TargetState::Sending(num);
```

Comment:
`Failed` is then be updated to `Sending` ?

Reply:
The target that is `Failed` with the lowest number of times is taken and updated to `Sending`. I can add a comment :+1: It's basically what JP described in issue #1684.